Go layout restructure

iwanbk opened the pull request:
Fixes #405 
- all types generated into `types` dir
- error if there are types which names only differ in case(lower/upper) of first character
- handlers are generated into `handlers` dir (same as python), each root endpoint into it's own package/dir. 

For example when we have two root endpoint `/users/` and `/members`. Directory structure of the generated server:

![structure](https://user-images.githubusercontent.com/977552/35435995-557c7418-02bf-11e8-887a-aa613af1e702.png)

serboctor commented on codegen/golang/resources.go:
```diff
@@ -86,14 +89,15 @@ func (gr *goResource) generateAPIImplementations(dir string) error {
 	}
+
+	apiDir := filepath.Join(dir, serverAPIDir, gr.PackageName)
 	if !apiFilePerMethod {
```
I think, like python, it shouldn't be an option to generate all the methods in the same file to make sure regeneration is always easy. @zaibon what do you think?

iwanbk replied:
i'll make `api-file-per-method` as default behavior, as as proposed (and not objected) at https://github.com/Jumpscale/go-raml/issues/405#issuecomment-356921853.

It'll then super easy to remove it

serboctor replied:
Yea, I thought it will be the default and only behaviour. It doesn't make sense to keep the option of generating all the methods in the same file. `api-file-per-method` should be removed from the command options and the golang generator should by default generate a file per method just like the python generation.

iwanbk replied:
make sense, i'll remove it.
`api-file-per-method` was added to keep backward compatibility.

serboctor commented on codegen/templates/golang/server_main_go.tmpl:
```diff
@@ -5,7 +5,9 @@ import (
 	"log"
```
Please add a comment at the top of this file stating whether it IS/IS NOT safe to edit.

iwanbk replied:
done

serboctor commented on codegen/templates/golang/server_resource_api_impl_go.tmpl:
```diff
@@ -3,7 +3,7 @@ package {{.PackageName}}
 
```
Please add a comment at the top of this file stating whether it IS/IS NOT safe to edit.

iwanbk replied:
fixed

serboctor commented on codegen/templates/golang/server_resources_api.tmpl:
```diff
@@ -3,7 +3,7 @@ package {{.PackageName}}
 
 import (
```
Please add a comment at the top of this file stating whether it IS/IS NOT safe to edit.

iwanbk replied:
fixed

serboctor commented:
If a file can be safely edited, we need a comment stating that. Check [this](https://github.com/Jumpscale/go-raml/blob/master/codegen/templates/python/server_resource_handler_flask.tmpl#L2) as an example.

iwanbk commented on codegen/golang/client.go:
```diff
@@ -66,6 +66,9 @@ func NewClient(apiDef *raml.APIDefinition, packageName, rootImportPath, targetDi
 
 // Generate generates all Go client files
 func (gc Client) Generate() error {
+	if err := checkDuplicatedTitleTypes(gc.apiDef); err != nil {
```
No, it is only needed by Go generator, because in Go we need first letter to be in uppercase for it to be exported.
Like described at https://github.com/Jumpscale/go-raml/issues/405#issuecomment-359663654

serboctor replied:
Actually in capnp we also camelcase the types. And in general, I dont think the user would create two types with different cases on purpose, so may be it would be a good thing to have it as a general validation step?

iwanbk replied:
Not sure about this. 

Why you made it camelcase?
Does capnp has technical reason like the Go has?

> And in general, I dont think the user would create two types with different cases on purpose

While it is true, we will violate raml spec for without good reason.

serboctor replied:
Yes, Capnp schema names have to start with an upper case. It has rules for the schema names and attributes.

iwanbk replied:
Oh, didn't know about it. I'll move it to `commons` then

iwanbk replied:
done

iwanbk commented:
> If a file can be safely edited, we need a comment stating that. Check this as an example.

fixed

serboctor commented on codegen/fixtures/security/deliveries_if.txt:
```diff
@@ -10,15 +10,15 @@ import (
 )
```
Please add a comment at the top of this file stating whether it IS/IS NOT safe to edit

serboctor commented on codegen/fixtures/server/user_api/users_if.txt:
```diff
@@ -9,14 +9,14 @@ import (
 )
```
Please add a comment at the top of this file stating whether it IS/IS NOT safe to edit

iwanbk replied:
oh, i missed this.

It actually already has, since long time ago.
```go
//This file is auto-generated by go-raml
//Do not edit this file by hand since it will be overwritten during the next generation
```
but yeah, it is better to make it consistent with other files

iwanbk commented:
@serboctor 
i've addressed your review, except about more checking to capnp https://github.com/Jumpscale/go-raml/pull/440#discussion_r164295785.
Because i don't think it is the right thing to do, yet i don't really know the capnp implementation.
